### PR TITLE
jwk #7: epoch manager

### PR DESCRIPTION
**Pull request by zjma**

### CODE REVIEW

File: crates/aptos-jwk-consensus/src/epoch_manager.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                self.consensus_key.clone(),
+                self.my_addr,
+                epoch_state.clone(),
+                Arc::new(qc_update_producer),
+                self.vtxn_pool.clone(),
+            );
+
+            let (jwk_event_tx, jwk_event_rx) = aptos_channel::new(QueueStyle::KLAST, 1, None);
+            self.jwk_updated_event_txs = Some(jwk_event_tx);
+            let (jwk_rpc_msg_tx, jwk_rpc_msg_rx) = aptos_channel::new(QueueStyle::FIFO, 100, None);
```
Comment (danielxiangzl) [inline crates/aptos-jwk-consensus/src/epoch_manager.rs:189]: Since it is per-issuer, this channel should also has size 1?

Comment (danielxiangzl) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:189]: Just checking, 100 is fine as well

Comment (zjma) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:189]: this channel is for epoch manager to forward network message to jwk manager, so not yet related to per-issuer logic?

Comment (danielxiangzl) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:189]: Is this the channel to forward peer request? Per issuer should have only 1 request per peer?

Comment (zjma) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:189]: yes it is to forward peer request. But there can be multiple concurrent requests e.g. peer 1 requesting google.com, peer 2 requesting facebook.com.

Comment (danielxiangzl) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:189]: Yes I see. you are right.

**Review by danielxiangzl — APPROVED** [review]

File: crates/aptos-jwk-consensus/src/epoch_manager.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        let epoch_state = Arc::new(EpochState {
+            epoch: payload.epoch(),
+            verifier: (&validator_set).into(),
+        });
+        self.epoch_state = Some(epoch_state.clone());
+        info!("[JWK] start_new_epoch: new_epoch={}", epoch_state.epoch);
+
+        let features = payload.get::<Features>().unwrap_or_default();
+
+        if features.is_enabled(FeatureFlag::JWK_CONSENSUS) {
```
Comment (danielxiangzl) [inline crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: Also need to check vtxn is enabled?

Comment (zjma) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: i still think we don't have to.
(Even for DKG i want to change back if we can).
1-to-1 mapping between flags and components would be ideal.

Comment (zekun000) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: what happens if vtxn is not enabled? just consensus wont' pull vtxn?

Comment (danielxiangzl) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: Then how do we ensure correctness? @zekun000 for discussion as well

File: types/src/contract_event.rs [hunk 3]
```diff
@@ -360,6 +361,24 @@ impl TryFrom<&ContractEvent> for DepositEvent {
...
+impl TryFrom<&ContractEvent> for ObservedJWKsUpdated {
+    type Error = Error;
+
+    fn try_from(event: &ContractEvent) -> Result<Self> {
+        match event {
+            ContractEvent::V1(_) => {
+                bail!("conversion to `ObservedJWKsUpdated` failed with wrong event version")
+            },
+            ContractEvent::V2(v2) => {
+                if v2.type_tag != TypeTag::Struct(Box::new(Self::struct_tag())) {
```
Comment (zekun000) [inline types/src/contract_event.rs:373]: can't make it lazy?

**Review by zekun000 — APPROVED** [review]

File: crates/aptos-jwk-consensus/src/epoch_manager.rs (hunk 2 above)
Comment (zjma) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: yes, consensus vtxn config controlls if consensus pull vtxns.
jwk feature controls if jwk vtxns are produced. (we have TxnGuard to ensure txns are dropped at the end of epoch)

Comment (danielxiangzl) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: nothing will break?

Comment (zjma) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: let's say nothing should break :)

File: types/src/contract_event.rs (hunk 3 above)
Comment (zjma) [reply types/src/contract_event.rs:373]: can try.

File: crates/aptos-jwk-consensus/src/epoch_manager.rs (hunk 2 above)
Comment (danielxiangzl) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: Can we add a smoke test for this? :)

Comment (zjma) [reply crates/aptos-jwk-consensus/src/epoch_manager.rs:160]: good idea.

**Pull request merged**